File: cpp/56_merge/merge.cpp
/*
 * @Date: 2023-08-27
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-08-27
 * @FilePath: /algorithm/cpp/56_merge/merge.cpp
 */

#include <algorithm>
#include <cassert>
#include <iostream>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> merge(vector<vector<int>>& intervals) {
    sort(intervals.begin(), intervals.end());
    vector<vector<int>> res{};
    for (size_t i = 0; i < intervals.size(); ++i) {
      int left = intervals[i][0], right = intervals[i][1];
      if (res.empty() || res.back()[1] < left) {
        res.push_back({left, right});
      } else {
        res.back()[1] = max(res.back()[1], right);
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<vector<int>>, vector<vector<int>>>> tests{
      {{{1, 3}, {2, 6}, {8, 10}, {15, 18}}, {{1, 6}, {8, 10}, {15, 18}}},
      {{{1, 4}, {4, 5}}, {{1, 5}}},
      {{{1, 4}, {0, 4}}, {{0, 4}}},
  };

  for (auto& [intervals, res] : tests) {
    assert(Solution().merge(intervals) == res);
  }
}

File: golang/56_merge/merge.go
/*
 * @Date: 2023-08-27
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-08-27
 * @FilePath: /algorithm/golang/56_merge/merge.go
 */

// Package main ...
package main

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func merge(intervals [][]int) (ans [][]int) {
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}

	sort.Slice(intervals, func(i, j int) bool { return intervals[i][0] < intervals[j][0] })
	for _, interval := range intervals {
		left := interval[0]
		right := interval[1]
		if len(ans) == 0 || ans[len(ans)-1][1] < left {
			ans = append(ans, []int{left, right})
		} else {
			ans[len(ans)-1][1] = max(ans[len(ans)-1][1], right)
		}
	}
	return
}

func main() {
	tests := []struct {
		intervals [][]int
		ans       [][]int
	}{
		{[][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}}, [][]int{{1, 6}, {8, 10}, {15, 18}}},
		{[][]int{{1, 4}, {4, 5}}, [][]int{{1, 5}}},
		{[][]int{{1, 4}, {0, 4}}, [][]int{{0, 4}}},
	}

	for _, item := range tests {
		assert.Equal(&testing.T{}, item.ans, merge(item.intervals))
	}
}
